File: Sputnik.cpp
/*

A simple Earth gravitation simulation.
Sputnik moving around the planet
(no effects to the planet applied).

Created and codded by Arseny Zolotarev (Tuesday, 18 of July, 2023)

© 2019-2023 Zolars
  
*/


#include <SFML/Graphics.hpp>
#include <iostream>
#include <vector>
#include <math.h>
#include <time.h>


// Including resource files path
#include "ResourcePath.hpp"


using namespace std;
using namespace sf;


// Framerate class
class FPS{
public:
    FPS(): mFrame(0), mFps(0) {}
    unsigned int getFPS() const { return mFps; }
    
    void update(){
        if(mClock.getElapsedTime().asSeconds() >= 1.f){ mFps = mFrame; mFrame = 0; mClock.restart(); }
        mFrame++;
    }
    
private:
    unsigned int mFrame, mFps;
    Clock mClock;
};


// 2D Vector structure
struct vec {
    long double x, y;
    double length(){
        return sqrt(x * x + y * y);
    }
};


// 2D object structure
struct Object {
    vec pos, vel = {1., 0.}, a = {0., 0.};
    double mass = 1.;
};


double dist(vec v1, vec v2){
    return sqrt((v1.x - v2.x)*(v1.x - v2.x) + (v1.y - v2.y)*(v1.y - v2.y));
}

double dist_S(vec v1, vec v2){
    return (v1.x - v2.x)*(v1.x - v2.x) + (v1.y - v2.y)*(v1.y - v2.y);
}


// Global varibales
double deltaTime = 0.01f, g = 9.81f;
float WIDTH = 800, HEIGHT = 800, H_WIDTH = WIDTH / 2, H_HEIGHT = HEIGHT / 2, ASPECT_RATIO = WIDTH / HEIGHT;


// Main function
int main() {
    // Creating environment
    Object planet, sputnik;
    
    planet.mass = 250;
    planet.pos = {0, 0};
    planet.vel = {0.025, -5};
    
    sputnik.mass = 1;
    sputnik.pos = {10, 0};
    sputnik.vel = {0, 5};
    
    // Trajectory vector
    VertexArray lines_s(LineStrip, 1);
    lines_s[0].position = Vector2f(sputnik.pos.x * 20 + H_WIDTH, H_HEIGHT - sputnik.pos.y * 20);
    int linesCount = 0;
    bool speedUp = false;
    
    // Creating window
    RenderWindow window(VideoMode(WIDTH, HEIGHT), "MIPH (visual)");
//    window.setFramerateLimit(100);
//    window.setVerticalSyncEnabled(true);
    
    FPS fps;
    Clock dTime;
    
    // Main cycle
    while (window.isOpen()){
        // User-input and Window events updating
        Event event;
        while (window.pollEvent(event)){
            if (event.type == Event::Closed)
                window.close();
            
            if (event.type == Event::KeyPressed){
                if (Keyboard::isKeyPressed(Keyboard::S))
                    speedUp ^= 1;
                
                if (Keyboard::isKeyPressed(Keyboard::R) || Keyboard::isKeyPressed(Keyboard::D)){
                    if (Keyboard::isKeyPressed(Keyboard::D)){ cout << "Write new time period: "; cin >> deltaTime; };

                    planet.mass = 250;
                    planet.pos = {0, 0};
                    planet.vel = {0.025, -5};
                    
                    sputnik.mass = 1;
                    sputnik.pos = {10, 0};
                    sputnik.vel = {0, 5};
                    
                    lines_s.clear();
                    linesCount = 0;
                    lines_s.append(Vector2f(sputnik.pos.x * 20 + H_WIDTH, H_HEIGHT - sputnik.pos.y * 20));
                    lines_s[linesCount].color = Color(100, 255, 100);
                }
            }
              
            // Mouse events
            if (event.type == Event::MouseButtonPressed){
                string filename = resourcePath() + "../../../screenshot.png";
                Texture texture;
                texture.create(window.getSize().x, window.getSize().y);
                texture.update(window);
                if (texture.copyToImage().saveToFile(filename)) printf("screenshot saved to %s\n", filename.c_str());
            }
        }
        

        // Drawing
        window.clear(Color(200, 200, 200));
        
        lines_s.append(sf::Vector2f(sputnik.pos.x * 20 + H_WIDTH, H_HEIGHT - sputnik.pos.y * 20));
        lines_s[++linesCount].color = Color(100, 255, 100);
        CircleShape ball(5, 10);
        CircleShape ball2(25, 25);

        ball.setPosition(sputnik.pos.x * 20 + H_WIDTH - 5, H_HEIGHT - 5 - sputnik.pos.y * 20);
        ball.setFillColor(Color(255, 100, 100));
        
        ball2.setPosition(planet.pos.x * 20 + H_WIDTH - 25, H_HEIGHT - 25 - planet.pos.y * 20);
        ball2.setFillColor(Color(255, 100, 100));
        
        window.draw(ball);
        window.draw(ball2);
        window.draw(lines_s);

        // Environments updating
        if (dTime.getElapsedTime().asMilliseconds() >= deltaTime * (speedUp ? 0 : 1000.)){
            sputnik.a = {
                planet.mass * sputnik.mass * -(sputnik.pos.x - planet.pos.x) / pow(dist_S(sputnik.pos, planet.pos), 1.5),
                planet.mass * sputnik.mass * -(sputnik.pos.y - planet.pos.y) / pow(dist_S(sputnik.pos, planet.pos), 1.5)};

            sputnik.vel.x += sputnik.a.x * deltaTime; sputnik.vel.y += sputnik.a.y * deltaTime;
            sputnik.pos.x += sputnik.vel.x / sputnik.mass * deltaTime; sputnik.pos.y += sputnik.vel.y / sputnik.mass * deltaTime;
            
            planet.a = {
                planet.mass * sputnik.mass * -(planet.pos.x - sputnik.pos.x) / pow(dist_S(sputnik.pos, planet.pos), 1.5),
                planet.mass * sputnik.mass * -(planet.pos.y - sputnik.pos.y) / pow(dist_S(sputnik.pos, planet.pos), 1.5)};

            planet.vel.x += planet.a.x * deltaTime; planet.vel.y += planet.a.y * deltaTime;
            planet.pos.x += planet.vel.x / planet.mass * deltaTime; planet.pos.y += planet.vel.y / planet.mass * deltaTime;
                        
            dTime.restart();
        }

        // Screen updating
        fps.update();
        window.setTitle("MIPH (visual) fps: " + to_string(fps.getFPS()));
        window.display();
    }

    return 0;
}

File: Sputnik.py
"""

A simple Earth gravitation simulation. 
Sputnik moving around the planet 
(no effects to the planet applied).

(And yes, it will be one-file)

Created and codded by Arseny Zolotarev (Tuesday, 18 of July, 2023)

© 2019-2023 Zolars
  
"""


import pygame as pg
import random as rand
import math
import sys


class Object:
    def __init__(self, pos=[0., 0.], vel=[1., 0.], mass=1.):
        self.pos = pos
        self.vel = vel
        self.mass = mass
        self.a = [0, 0]


# global references
try:
    SIZE = WIDTH, HEIGHT = int(sys.argv[1]), int(sys.argv[2])
except IndexError:
    SIZE = WIDTH, HEIGHT = 800, 800
H_SIZE = H_WIDTH, H_HEIGHT = WIDTH / 2, HEIGHT / 2
l_press, r_press = False, False
DELTA_TIME = 0.01
g = 9.81


def transform_coords(position):
    return [position[0] * 20 + H_WIDTH, H_HEIGHT - position[1] * 20]


def length(vec1, vec2):
    return math.sqrt((vec1[0] - vec2[0])*(vec1[0] - vec2[0]) + (vec1[1] - vec2[1])*(vec1[1] - vec2[1]))


def length_S(vec1, vec2):
    return math.sqrt((vec1[0] - vec2[0])*(vec1[0] - vec2[0]) + (vec1[1] - vec2[1])*(vec1[1] - vec2[1]))


def update_state(obj1, obj2):
    # acceleration
    obj2.a[0] = obj1.mass * obj2.mass * obj2.pos[0] / -length_S(obj1.pos, obj2.pos)
    obj2.a[1] = obj1.mass * obj2.mass * obj2.pos[1] / -length_S(obj1.pos, obj2.pos)

    obj1.a[0] = obj1.mass * obj2.mass * obj1.pos[0] / -length_S(obj1.pos, obj2.pos)
    obj1.a[1] = obj1.mass * obj2.mass * obj1.pos[1] / -length_S(obj1.pos, obj2.pos)

    # sputnik
    obj2.vel[0] += obj2.a[0] * DELTA_TIME
    obj2.vel[1] += obj2.a[1] * DELTA_TIME
    obj2.pos[0] += obj2.vel[0] * DELTA_TIME
    obj2.pos[1] += obj2.vel[1] * DELTA_TIME

    # planet
    obj1.vel[0] += obj1.a[0] * DELTA_TIME
    obj1.vel[1] += obj1.a[1] * DELTA_TIME
    obj1.pos[0] += obj1.vel[0] * DELTA_TIME
    obj1.pos[1] += obj1.vel[1] * DELTA_TIME


if __name__ == "__main__":
    # creating a pygame window
    screen = pg.display.set_mode(SIZE, vsync=1)
    clock = pg.time.Clock()

    # creating the game logic / environment sample
    planet = Object(mass=5, vel=[0., -0.25], pos=[0, 0])
    sputnik = Object(mass=1, vel=[0, 6.75], pos=[10, 0])

    upd_ticks = pg.time.get_ticks()
    speedup = False
    trajectory_s = [transform_coords(sputnik.pos), transform_coords(sputnik.pos)]
    trajectory_p = [transform_coords(planet.pos), transform_coords(planet.pos)]

    # main cycle
    while True:
        screen.fill((200, 200, 200))

        # checking for keyboard, window, mouse inputs or events
        keys = pg.key.get_pressed()
        for event in pg.event.get():
            if event.type == pg.QUIT:
                exit()
            if event.type == pg.KEYDOWN:
                if event.key == pg.K_c:
                    trajectory_s = [transform_coords(sputnik.pos), transform_coords(sputnik.pos)]
                    trajectory_p = [transform_coords(planet.pos), transform_coords(planet.pos)]

                if event.key == pg.K_s:
                    speedup = not speedup

                # restarting the simulation environment
                if event.key == pg.K_d or event.key == pg.K_r:
                    if event.key == pg.K_d:
                        DELTA_TIME = float(input("Write new time period: "))

                    planet = Object(mass=5, vel=[0., -0.25], pos=[0, 0])
                    sputnik = Object(mass=1, vel=[0, 6.75], pos=[10, 0])

                    trajectory_s = [transform_coords(sputnik.pos), transform_coords(sputnik.pos)]
                    trajectory_p = [transform_coords(planet.pos), transform_coords(planet.pos)]

        # game Assets/UI/elements drawing
        pg.draw.circle(screen, (255, 100, 100), transform_coords(planet.pos), 25)
        pg.draw.circle(screen, (255, 100, 100), transform_coords(sputnik.pos), 5)
        # trajectories
        pg.draw.lines(screen, (100, 255, 100), False, trajectory_p, 3)
        pg.draw.lines(screen, (100, 255, 100), False, trajectory_s, 3)

        # game environment updating (with vertical synchronization)
        if 0 <= pg.time.get_ticks() - upd_ticks - (DELTA_TIME * (speedup and 100 or 1000)):
            upd_ticks = pg.time.get_ticks()

            # calling for game environment to update
            update_state(planet, sputnik)
            trajectory_p.append(transform_coords(planet.pos))
            trajectory_s.append(transform_coords(sputnik.pos))

        pg.display.set_caption("$~MIPH ~fps: " + str(round(clock.get_fps(), 2)))

        pg.display.flip()
        clock.tick()
